torch.tensor can infer complex dtype now

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33507 Gradcheck for complex
* **#33361 torch.tensor can infer complex dtype now**
* #33268 Added tensor.is_complex(), is_complex and dtype.is_complex py binding, tensor printing, and dixed the scalar type returned for complex float

Differential Revision: [D19943477](https://our.internmc.facebook.com/intern/diff/D19943477)

anjali411 committed:
torch.tensor can infer complex dtype now

[ghstack-poisoned]

anjali411 committed:
Update on "torch.tensor can infer complex dtype now"




[ghstack-poisoned]

anjali411 committed:
Update on "torch.tensor can infer complex dtype now"




[ghstack-poisoned]

anjali411 committed:
Update on "torch.tensor can infer complex dtype now"




[ghstack-poisoned]

anjali411 committed:
Update on "torch.tensor can infer complex dtype now"


Differential Revision: [D19943477](https://our.internmc.facebook.com/intern/diff/D19943477)

[ghstack-poisoned]

anjali411 committed:
Update on "torch.tensor can infer complex dtype now"


Differential Revision: [D19943477](https://our.internmc.facebook.com/intern/diff/D19943477)

[ghstack-poisoned]

anjali411 committed:
Update on "torch.tensor can infer complex dtype now"


Differential Revision: [D19943477](https://our.internmc.facebook.com/intern/diff/D19943477)

[ghstack-poisoned]

ezyang commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -169,6 +169,9 @@ ScalarType infer_scalar_type(PyObject *obj) {
     return ScalarType::Bool;
   }
+  if (PyComplex_Check(obj)) {
+    return ScalarType::ComplexDouble;
```
However, I am wondering; above we have `torch::tensors::get_default_scalar_type()`, so I wonder if complex should switch between ComplexDouble and ComplexFloat depending on the default scalar type. Something to think about.

anjali411 replied:
I think that's because those float and double exist in std library, so we can just easily infer the scalar types accordingly.

ezyang replied:
I'm not sure what you mean by "exist in the std library"; `get_default_scalar_type` doesn't have much to do with `std`.

anjali411 committed:
Update on "torch.tensor can infer complex dtype now"


Differential Revision: [D19943477](https://our.internmc.facebook.com/intern/diff/D19943477)

[ghstack-poisoned]

anjali411 committed:
Update on "torch.tensor can infer complex dtype now"


Differential Revision: [D19943477](https://our.internmc.facebook.com/intern/diff/D19943477)

[ghstack-poisoned]

anjali411 committed:
Update on "torch.tensor can infer complex dtype now"


Differential Revision: [D19943477](https://our.internmc.facebook.com/intern/diff/D19943477)

[ghstack-poisoned]

anjali411 committed:
Update on "torch.tensor can infer complex dtype now"


Differential Revision: [D19943477](https://our.internmc.facebook.com/intern/diff/D19943477)

[ghstack-poisoned]

anjali411 committed:
Update on "torch.tensor can infer complex dtype now"


Differential Revision: [D19943477](https://our.internmc.facebook.com/intern/diff/D19943477)

[ghstack-poisoned]

anjali411 committed:
Update on "torch.tensor can infer complex dtype now"


Differential Revision: [D19943477](https://our.internmc.facebook.com/intern/diff/D19943477)

[ghstack-poisoned]

anjali411 committed:
Update on "torch.tensor can infer complex dtype now"


Differential Revision: [D19943477](https://our.internmc.facebook.com/intern/diff/D19943477)

[ghstack-poisoned]
